libr/include/r_util/r_annotated_code.h

```diff
@@ -56,6 +58,14 @@ typedef struct r_code_annotation_t {
 			char *name;
 			ut64 offset; // offset of the function with the name specified
 		} function_name;
+
+		struct {
+			ut64 offset;
+		} global_variable;
+
+		struct {
+			ut64 offset;
+		} constant_variable;
```

Comment:
Instead of having separate enum variants for function, global variable and constant I suggest merging them into single one called reference. In many cases using it don't care if it's a reference to function, global variable or read only global variable, just the fact that it points to some fixed memory address. That way the code using it can be simplified. You can keep the separate RCodeAnnotationType values so that they can be differentiated if necessary.

Reply:
@karliss I think it's a good idea to merge them. Functions have `char *name`, so when the type is not function, it will point to `nullptr`.

Reply:
@karliss Please have a look now.

Reply:
I don't think it's necessary to have two nested type specifiers unless the structure gets significantly more complicated, but that seems unlikely. If you decide to keep both of them don't reuse the same enum type for both. If you keep single type field there should be a comment explaining what RCodeAnnotationType use the reference field.

Reply:
It is to differentiate between function annotations and variable annotations. It's useful for printing JSON also.

> If you decide to keep both of them don't reuse the same enum type for both.

I didn't understand this. I will add a comment that explains what RCodeAnnotationType values are used by `annotation.reference.type`.

Reply:
You can still differentiate between the 3 if you store R_CODE_ANNOTATION_TYPE_FUNCTION_NAME, R_CODE_ANNOTATION_TYPE_GLOBAL_VARIABLE, R_CODE_ANNOTATION_TYPE_CONSTANT_VARIABLE directly in the top level type field without having a second type field inside union and R_CODE_ANNOTATION_TYPE_REFERENCE.

Currently you have two type fields one directly inside the RAnnotatedCode and one within the union reference case.  In your current design top one will always be _OFFSET, _SYNTAX_HIGHLIGHT or _REFERENCE but never FUNCTION_NAME or VARIABLE. The type inside reference struct will always be one of the last 3 but never one of the first 3. If it is used that way you should define two enums instead of using single type for two different purpose.

Out of the two approaches described above I recommend the first one.

Reply:
I'll go with the first one:)

Reply:
Is there need to keep R_CODE_ANNOTATION_TYPE_REFERENCE type instead of using the more specific ones? Also the current situation where multiple annotation types use the same enum field was for which I suggested to add a comment. Explaining the mapping.

Reply:
Removed R_CODE_ANNOTATION_TYPE_REFERENCE and explained the mapping with a comment.